### ISSUE TITLE
`fish_preexec` current directory not consistent?

### ISSUE THREAD

**Issue by aequitas** [fish-shell/fish-shell#6465]
```
~ $ fish --version 
fish, version 3.0.2
```

```
~ $ uname -a 
Darwin Big-McIntosh 18.7.0 Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64 x86_64
```

macOS Terminal.app
```
~ $ echo $TERM
xterm-256color
```

I tried using the `sh -c 'env HOME=$(mktemp -d) fish'` command, problem persists.

I have two systems with same fish version (and same direnv 2.20.1 and macOS version 10.14.6). On both systems I have enabled a Direnv hook via `direnv hook fish | source`. On one system Direnv works as expected, ie: when I enter a directory the `.envrc` file is sourced. On the other (newly installed) system Direnv 'lags' one prompt behind, ie: when I enter the directory the `.envrc` file is not noticed, but when I hit Enter to force a new prompt it gets noticed and sourced.

Debugging leads me to believe this is because the `fish_preexec` event that is used by the Direnv hook [0] has different behaviour on both systems.

- On the working system the function is executed in the newly cd'd to directory. 
- On the not-working system it is exectuted in the previous directory. 
- When changing the event from `fish_preexec` to `fish_prompt` the behaviour is as expected, the function is executed in the newly cd'd to directory. 

See https://asciinema.org/a/jl7G5qNPNZ0qAVRoTFHHiOMXH

[0]
```
function __direnv_export_eval --on-event fish_preexec
        "/usr/local/bin/direnv" export fish | source;
end
```

**Comment by faho** [comment]
Please list all preexec handlers with `functions --handlers`, and confirm that the `cd` function is the same on both.

I'm pretty sure it's not about preexec being triggered before/after changing $PWD depending on the system (btw why *pre*exec? If you want `cd somedir/` to source `somedir/.direnv` wouldn't you want *post*exec?)

**Comment by aequitas** [comment]
Thanks for the quick reply. I made a bad assumption, on the working system the event was actually `fish_prompt`, which is why it worked there.

I've looked into it more on the Direnv side and it seems the hook would previously use `fish_prompt` where since https://github.com/direnv/direnv/pull/512 it uses `fish_preexec`. Explaining why it worked on my old system because the hook there was generated by a older version of the software. I'll make this an issue at Direnv as it seems that where the problem stems from.

`fish_postexec` might be what is needed instead of `fish_preexec` or `fish_prompt`. Thanks.
